### ISSUE TITLE
Support for compiling and linking non-C objects (such as Fortran)

### ISSUE THREAD

**Issue by berke** [ocaml/dune#803]
Hi,

I have some F90 code in my project that I'd like to be built and linked with the rest.  Is there a way of doing this with Dune?  If there is it's not obvious, neither from the documentation nor from a quick glance at the code :)  Of course for C it's easy.

For reference with Oasis & Ocamlbuild I use an .f90-to-.o rule in myocamlbuild.ml and then I cheat by telling Oasis that it's C code (i.e. I say CSources: foo.c when in reality the file is foo.f90).

**Comment by Chris00** [comment]
@berke You may want to have a look to https://github.com/Chris00/L-BFGS-ocaml (the FORTRAN code is in [src/Lbfgsb.3.0](https://github.com/Chris00/L-BFGS-ocaml/tree/master/src/Lbfgsb.3.0)).
I guess a FORTRAN package detecting the FORTRAN compiler and providing new Dune rules would be nice...

**Comment by berke** [comment]
@diml Yes #788 might do the trick, and @Chris00 that's the perfect example.  Will try that!   Many thanks.

**Comment by rgrinberg** [comment]
@nojb @snowleopard where are we on this front? As far as I can tell, it's still not possible to add arbitrary object files.

**Comment by snowleopard** [comment]
@rgrinberg I think we still don't have a syntax for declaring custom-built object files. However, we do have a way to depend on custom-built [foreign archives](https://dune.readthedocs.io/en/stable/concepts.html#foreign-archives), so I guess one can use it by adding a custom `ar` rule.

Adding a special syntax for linking non-C objects doesn't seem complicated. I'm happy implement this once I plough through the rest of my todo list, or review a PR if someone beats me to it.
